designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,469 @@
+- (no RFC yet) We can support ES modules for shareable configs, plugins, and custom parsers.
+
+##### Move the `usedDeprecatedRules` property
+
+The returned object of `CLIEngine#executeOnFiles()` has the `usedDeprecatedRules` property that includes the deprecated rule IDs which the linting used.
+
+But the location is problematic because it requires the plugin uniqueness in spanning all files. Therefore, this RFC moves the `usedDeprecatedRules` property to each lint result.
+
+```js
+const { LinterShell } = require("eslint")
+const linter = new LinterShell()
+
+for (const result of await linter.lintFiles(patterns)) {
+  console.log(result.usedDeprecatedRules)
+}
+```
+
+As a side-effect, formatters gets the capability to print the used deprecated rules. Previously, ESLint has not passed the returned object to formatters, so the formatters could not print used deprecated rules. After this proposal, each lint result has the `usedDeprecatedRules` property and the formatters receive those.
+
+##### Write cache safely (best effort)
+
+Because this method updates the cache file, it will break the cache file if called multiple times in parallel. To prevent that, this method doesn't write the cache file if the cache file has been updated since this method read.
+
+This method does this check with the best effort (e.g., check `mtime` of the cache file) because Node.js doesn't provide the way that reads/writes a file exclusively.
+
+If the cache file was broken, this method should ignore the cache file and does lint.
+
+##### Abort linting
+
+This proposal doesn't provide the way to abort linting because we cannot abort linting at this time (the method does linting synchronously internally). We can discuss it along with parallel linting (I'm guessing we can use [AbortSignal] that is the Web Standard. Passing it as `options.signal`).
+
+#### ● The `lintText()` method
+
+##### Parameters
+
+| Name                  | Type      | Description                                                      |
+| :-------------------- | :-------- | :--------------------------------------------------------------- |
+| `code`                | `string`  | The source code to lint.                                         |
+| `options`             | `Object`  | Optional. The options.                                           |
+| `options.filePath`    | `string`  | Optional. The path to the file of the source code.               |
+| `options.warnIgnored` | `boolean` | Optional. If `true`, it warns the `filePath` is an ignored path. |
+
+##### Return Value
+
+This method returns a Promise object that will be fulfilled with the array of lint results.
+
+##### Description
+
+This method corresponds to `CLIEngine#executeOnText()`.
+
+Because the returned object of `CLIEngine#executeOnText()` method is the same type as the `CLIEngine#executeOnFiles()` method, the `LinterShell` class inherits that mannar. Therefore, the returned value is an array that contains one result.
```

Comment:
Small wording improvement/clarification:

```suggestion
Because the returned object of `CLIEngine#executeOnText()` method is the same type as the `CLIEngine#executeOnFiles()` method, the `LinterShell` class also returns the same type. Therefore, the returned value is an array that contains one result.
```